src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -96,15 +97,23 @@ private BoundExpression RewriteStringConcatenation(SyntaxNode syntax, BinaryOper
                 case 2:
                     var left = leftFlattened[0];
                     var right = leftFlattened[1];
-                    result = RewriteStringConcatenationTwoExprs(syntax, left, right);
+
+                    if (!TryRewriteStringConcatenationWithSpanBasedConcat(syntax, leftFlattened.ToImmutable(), out result))
```

Comment:
>leftFlattened.ToImmutable() [](http://example.com/codeflow?start=82&length=27)

It might be better to avoid allocating an array just to try something out. Could we simply pass the builder and allocate only for success case and only what we need?